### DISCUSSION TITLE
Improve display of total-line-numbers in the statusline

### DISCUSSION THREAD

**Discussion by antoyo** [helix-editor/helix#10397] [Ideas]
Hi.
Currently, the display for current line and total line numbers in the statusline is separate.
I'd prefer if it could be combined.
For instance, to have something like this:
```
5/15:2

(current line)/(total line numbers):(column)
```
instead of:
```
5:2 15
```
I'm also open to other ideas, but I like having `(current line)/(total line numbers)`.
Thanks.

**Comment by tysteiman** [comment]
@antoyo Given that the total-line-numbers element is rendered separately from the position element:

https://github.com/helix-editor/helix/blob/master/helix-term/src/ui/statusline.rs#L154

Shifting around the elements of the statusline would break that formatting. For example, if you move an element (in this case selection) _before_ total-line-numbers, it currently looks like this:

![image](https://user-images.githubusercontent.com/11522852/206356282-b6ca7efa-6440-494b-894a-3c9a74f3203a.png)

With the 14 on the right being the total line numbers.

total-line-numbers would either have to be baked into position or we would have to place things in specific orders depending on if they're set, neither of those sound like great options for customization.

Other than that we would be looking at the bigger picture on how to let users customize the statusline on a deeper level to include a format string of some sort instead of just an array of elements.

Thoughts? Hopefully I didn't misunderstand or overthink it.

**Comment by austinletson** [comment]
@antoyo @tysteiman What do you think about allowing the user to setting a format string containing the variables `col`, `row`, `total-line-number` like this:
```
[editor.statusline]
position_format = "{row}/({total-line-numbers}:{col}"
```

This way the user could use any formatting they want including those three variables and chars like `:` and `/`. This seems like an easy way to make other parts of the status line more customizable as well but I don't know what the roadmap is for such customization or if it is even desired in the long run. Allowing custom formatting for these three variables seems like a good starting point to test out a strategy that could be abstracted to other parts of the line.



Just to give an example, here is a quick change I made to the `render_position` function in `statusline.rs:332`. This implementation doesn't work, because the `format!` macro that is used throughout `statusline.rs` requires a string literal, but there seem to be several simple libraries for dynamically formatting strings with a variable format string. These might already be used elsewhere in the code base. I am new to the code base.

```
fn render_position<F>(context: &mut RenderContext, write: F)
where
    F: Fn(&mut RenderContext, String, Option<Style>) + Copy,
{
    let position = get_position(context);
    let config = context.editor.config();

    // position_format contains a standard string formatting with named arguments for row and col
    let position_format = &config.statusline.position_format;

    // Since position.row and position.col are zero-indexed we need to ad one to have human readable values
    let adjusted_row = position.row + 1;
    let adjusted_col = position.col + 1;
    write(
        context,
        format!(
            // KEY LINE
            position_format.to_string(),
            row = adjusted_row,
            col = adjusted_col
        ),
        None,
    );
}
```
 Also here is the minor change to `editor.rs:266` that would make the above `&config.statusline.position_format` work. The key here is the addition of the `position_format` member to the `StatusLineConfig` struct.

```
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case", default, deny_unknown_fields)]
pub struct StatusLineConfig {
    pub left: Vec<StatusLineElement>,
    pub center: Vec<StatusLineElement>,
    pub right: Vec<StatusLineElement>,
    pub separator: String,
    // KEY LINE
    pub position_format: String,
    pub mode: ModeConfig,
}

impl Default for StatusLineConfig {
    fn default() -> Self {
        use StatusLineElement as E;

        Self {
            left: vec![E::Mode, E::Spinner, E::FileName],
            center: vec![],
            right: vec![E::Diagnostics, E::Selections, E::Position, E::FileEncoding],
            separator: String::from("│"),
            // KEY LINE
            position_format: String::from("{row}:{col}"),
            mode: ModeConfig::default(),
        }
    }
}
```

Edit: also here is the branch with these minor changes: https://github.com/austinletson/helix/tree/improve-display-of-total-line-numbers-in-status-line

**Comment by antoyo** [comment]
I would be OK with the format string.

I'm not sure how much users really want that level of customization. If they don't, perhaps a simpler solution would be to create a new statusline element containing both the line/col and total number of lines.
